File: modules/expo-sticker-runtime/android/src/main/java/expo/modules/stickerruntime/ExpoStickerRuntimeModule.kt
package expo.modules.stickerruntime

import android.app.ActivityManager
import android.os.Build
import android.os.StatFs
import android.util.Log
import ai.onnxruntime.OrtSession
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.delay
import java.io.File
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicReference

private data class ActiveGeneration(
  val cancelled: AtomicBoolean = AtomicBoolean(false),
  val runOptions: AtomicReference<OrtSession.RunOptions?> = AtomicReference(null),
)

class ExpoStickerRuntimeModule : Module() {
  private val activeRequests = ConcurrentHashMap<String, ActiveGeneration>()
  private val context get() = requireNotNull(appContext.reactContext)
  private val bundles by lazy { ModelBundleManager(context) }
  private val inference by lazy {
    StickerInferenceEngine(
      context,
      requireNotNull(runtimeDecision().sessionPolicy) { "RUNTIME_UNAVAILABLE" },
    )
  }

  override fun definition() = ModuleDefinition {
    Name("ExpoStickerRuntime")
    Events("onModelDownloadProgress", "onGenerationProgress")

    AsyncFunction("getCapabilities").SuspendBody<Map<String, Any?>> {
      capabilities()
    }

    AsyncFunction("getModelBundleState").SuspendBody<Map<String, Any?>> {
      bundleState()
    }

    AsyncFunction("startModelDownload").SuspendBody<Map<String, Any?>> {
      val manifest = bundles.loadManifest()
      if (!bundles.hasEnoughStorage(manifest)) {
        return@SuspendBody state("failed", "INSUFFICIENT_STORAGE", "Not enough free storage")
      }
      bundles.cancel(manifest)
      bundles.enqueue(manifest)
      while (true) {
        val (downloaded, total) = bundles.progress(manifest)
        sendEvent("onModelDownloadProgress", mapOf(
          "phase" to "downloading",
          "downloadedBytes" to downloaded,
          "totalBytes" to total,
        ))
        if (downloaded >= total) break
        delay(500)
      }
      sendEvent("onModelDownloadProgress", mapOf(
        "phase" to "verifying",
        "downloadedBytes" to manifest.artifactBytes,
        "totalBytes" to manifest.artifactBytes,
      ))
      bundles.promote(manifest)
      state("ready")
    }

    AsyncFunction("installLocalModel").SuspendBody<Map<String, Any?>> {
      if (!BuildConfig.DEBUG) {
        return@SuspendBody state(
          "failed",
          "LOCAL_MODEL_NOT_STAGED",
          "Local model import is available only in debug builds",
        )
      }
      val manifest = bundles.loadManifest()
      sendEvent("onModelDownloadProgress", mapOf(
        "phase" to "verifying",
        "downloadedBytes" to 0L,
        "totalBytes" to manifest.artifactBytes,
      ))
      try {
        bundles.installLocal(manifest)
        sendEvent("onModelDownloadProgress", mapOf(
          "phase" to "verifying",
          "downloadedBytes" to manifest.artifactBytes,
          "totalBytes" to manifest.artifactBytes,
        ))
        state("ready")
      } catch (error: LocalModelImportException) {
        Log.e("ExpoStickerRuntime", error.message, error)
        state("failed", error.code, error.message ?: error.code)
      } catch (error: Throwable) {
        state("failed", "MODEL_PROMOTION_FAILED", error.message ?: "Local model import failed")
      }
    }

    AsyncFunction("cancelModelDownload").SuspendBody<Map<String, Any?>> {
      runCatching { bundles.cancel(bundles.loadManifest()) }
      state("missing")
    }

    AsyncFunction("prepareModel").SuspendBody<Map<String, Any?>, Map<String, Any?>> { _ ->
      val manifest = bundles.loadManifest()
      if (!bundles.isReady(manifest)) {
        mapOf(
          "contractVersion" to "1.0",
          "modelId" to manifest.modelId,
          "modelVersion" to manifest.modelVersion,
          "ready" to false,
          "errorCode" to "MODEL_NOT_AVAILABLE",
        )
      } else {
        mapOf(
          "contractVersion" to "1.0",
          "modelId" to manifest.modelId,
          "modelVersion" to manifest.modelVersion,
          "ready" to true,
        )
      }
    }

    AsyncFunction("generate").SuspendBody<Map<String, Any?>, Map<String, Any?>> { request ->
      val requestId = request["requestId"] as? String ?: error("requestId is required")
      val active = ActiveGeneration()
      check(activeRequests.putIfAbsent(requestId, active) == null) {
        "GENERATION_BUSY"
      }
      val startedAt = System.currentTimeMillis()
      val sequence = AtomicInteger(0)
      try {
        val manifest = bundles.loadManifest()
        check(bundles.isReady(manifest)) { "MODEL_NOT_AVAILABLE" }
        inference.generate(
          request,
          active.cancelled,
          active.runOptions::set,
        ) { stage, progress ->
          sendEvent(
            "onGenerationProgress",
            mapOf(
              "contractVersion" to "1.0",
              "requestId" to requestId,
              "sequence" to sequence.incrementAndGet(),
              "stage" to stage,
              "stageProgress" to progress,
              "elapsedMs" to System.currentTimeMillis() - startedAt,
            ),
          )
        }
      } catch (error: Throwable) {
        File(context.cacheDir, "generated-stickers/$requestId.png").delete()
        if (active.cancelled.get()) {
          throw IllegalStateException("GENERATION_CANCELLED", error)
        }
        Log.e("ExpoStickerRuntime", "Generation $requestId failed: ${error.message}", error)
        throw error
      } finally {
        activeRequests.remove(requestId)
      }
    }

    AsyncFunction("cancel").SuspendBody<Map<String, Any?>, String> { requestId ->
      val active = activeRequests[requestId]
      val accepted = active?.cancelled?.compareAndSet(false, true) ?: false
      if (accepted) {
        val runOptions = active?.runOptions?.get()
        Log.i(
          "ExpoStickerRuntime",
          RuntimeDiagnostics.cancellationAccepted(requestId, runOptions != null),
        )
        runCatching { runOptions?.setTerminate(true) }
      }
      mapOf(
        "accepted" to accepted,
        "outcome" to if (accepted) "cancellation_requested" else "not_found",
      )
    }
  }

  private fun capabilities(): Map<String, Any?> {
    val decision = runtimeDecision()
    val storage = StatFs(context.filesDir.absolutePath).availableBytes
    return if (decision.supported) {
      mapOf(
        "supported" to true,
        "adapterId" to "expo-sticker-runtime-onnx",
        "totalMemoryClassMb" to totalMemoryMb(),
        "deviceKind" to decision.deviceKind,
        "architecture" to decision.architecture,
        "availableDelegates" to listOf("NNAPI"),
        "selectedDelegate" to if (decision.sessionPolicy == SessionPolicy.EMULATOR_NNAPI) {
          "NNAPI_REFERENCE"
        } else {
          "NNAPI"
        },
        "runtimeVersion" to "1.27.0",
        "availableStorageBytes" to storage,
      )
    } else {
      mapOf(
        "supported" to false,
        "reasonCode" to decision.reasonCode,
        "deviceKind" to decision.deviceKind,
        "architecture" to decision.architecture,
      )
    }
  }

  private fun runtimeDecision(): RuntimeDecision = RuntimePolicy.evaluate(
    debug = BuildConfig.DEBUG,
    emulator = isEmulator(),
    supportedAbis = Build.SUPPORTED_ABIS.toList(),
    totalMemoryMb = totalMemoryMb(),
  )

  private fun totalMemoryMb(): Long {
    val activityManager = context.getSystemService(ActivityManager::class.java)
    val memory = ActivityManager.MemoryInfo().also(activityManager::getMemoryInfo)
    return memory.totalMem / (1024 * 1024)
  }

  private fun isEmulator(): Boolean =
    Build.FINGERPRINT.startsWith("generic") ||
      Build.FINGERPRINT.contains("emulator") ||
      Build.MODEL.contains("Emulator") ||
      Build.MODEL.contains("Android SDK built for") ||
      Build.PRODUCT.contains("sdk_gphone") ||
      Build.HARDWARE == "ranchu" ||
      Build.HARDWARE == "goldfish"

  private fun bundleState(): Map<String, Any?> = runCatching {
    val manifest = bundles.loadManifest()
    if (bundles.isReady(manifest)) state("ready") else state("missing")
  }.getOrElse {
    state("failed", "MODEL_MANIFEST_MISSING", it.message ?: "Model manifest is unavailable")
  }

  private fun state(status: String, code: String? = null, message: String? = null): Map<String, Any?> =
    mapOf(
      "status" to status,
      "modelId" to "lcm-sd15-chibi",
      "modelVersion" to "1.0.1",
      "downloadedBytes" to if (status == "ready") runCatching {
        bundles.loadManifest().artifactBytes
      }.getOrDefault(0L) else 0L,
      "totalBytes" to runCatching { bundles.loadManifest().artifactBytes }.getOrDefault(0L),
      "errorCode" to code,
      "message" to message,
    )
}

File: modules/expo-sticker-runtime/ios/ExpoStickerRuntimeModule.swift
import Darwin
import ExpoModulesCore
import Foundation
import UIKit

public final class ExpoStickerRuntimeModule: Module {
  private let bundles = ModelBundleManager()
  private let generationLock = NSLock()
  private var activeRequest: (id: String, cancellation: GenerationCancellation)?

  public func definition() -> ModuleDefinition {
    Name("ExpoStickerRuntime")
    Events("onModelDownloadProgress", "onGenerationProgress")

    AsyncFunction("getCapabilities") { () -> [String: Any] in
      self.capabilities()
    }

    AsyncFunction("getModelBundleState") { () -> [String: Any] in
      self.bundleState()
    }

    AsyncFunction("startModelDownload") { () async -> [String: Any] in
      do {
        let manifest = try self.bundles.loadManifest()
        try await self.bundles.downloadAndPromote(manifest) { phase, downloaded, total in
          self.sendEvent("onModelDownloadProgress", [
            "phase": phase,
            "downloadedBytes": downloaded,
            "totalBytes": total,
          ])
        }
        return self.state("ready", manifest: manifest)
      } catch {
        return self.failedState(error)
      }
    }

    AsyncFunction("installLocalModel") { () -> [String: Any] in
      #if DEBUG
      do {
        let manifest = try self.bundles.loadManifest()
        let documents = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let archive = documents.appendingPathComponent(
          "lcm-sd15-chibi-coreml-ios-v1.0.1.zip"
        )
        guard FileManager.default.fileExists(atPath: archive.path) else {
          return self.state(
            "failed",
            manifest: manifest,
            errorCode: "LOCAL_MODEL_NOT_STAGED"
          )
        }
        try self.bundles.installLocal(archive, manifest: manifest) { phase, downloaded, total in
          self.sendEvent("onModelDownloadProgress", [
            "phase": phase,
            "downloadedBytes": downloaded,
            "totalBytes": total,
          ])
        }
        return self.state("ready", manifest: manifest)
      } catch {
        return self.failedState(error)
      }
      #else
      return self.state("failed", errorCode: "LOCAL_MODEL_NOT_STAGED")
      #endif
    }

    AsyncFunction("cancelModelDownload") { () -> [String: Any] in
      self.bundles.cancelDownload()
      return self.state("missing")
    }

    AsyncFunction("prepareModel") { (request: [String: Any]) -> [String: Any] in
      let requestedId = request["modelId"] as? String ?? ""
      let requestedVersion = request["modelVersion"] as? String ?? ""
      guard let manifest = try? self.bundles.loadManifest(),
        requestedId == manifest.modelId,
        ["1.0.1", manifest.modelVersion].contains(requestedVersion),
        self.bundles.isReady(manifest)
      else {
        return [
          "contractVersion": "1.0",
          "modelId": requestedId,
          "modelVersion": requestedVersion,
          "ready": false,
          "errorCode": "MODEL_NOT_AVAILABLE",
        ]
      }
      return [
        "contractVersion": "1.0",
        "modelId": manifest.modelId,
        "modelVersion": manifest.modelVersion,
        "ready": true,
      ]
    }

    AsyncFunction("generate") { (request: [String: Any]) throws -> [String: Any] in
      let parsed = try self.parseGenerationRequest(request)
      let cancellation = GenerationCancellation()
      guard self.register(parsed.requestId, cancellation: cancellation) else {
        throw NSError(
          domain: "GENERATION_BUSY",
          code: 1,
          userInfo: [NSLocalizedDescriptionKey: "GENERATION_BUSY"]
        )
      }
      let started = Date()
      var sequence = 0
      var terminal = false
      defer { self.unregister(parsed.requestId) }
      do {
        let manifest = try self.bundles.loadManifest()
        guard self.bundles.isReady(manifest) else {
          throw StickerRuntimeError.modelNotAvailable
        }
        let engine = StickerInferenceEngine(
          modelRoot: self.bundles.readyURL,
          cacheDirectory: FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0]
        )
        let output = try engine.generate(parsed, cancellation: cancellation) { stage, value in
          guard !terminal else { return }
          sequence += 1
          self.sendEvent("onGenerationProgress", [
            "contractVersion": "1.0",
            "requestId": parsed.requestId,
            "sequence": sequence,
            "stage": stage,
            "stageProgress": value,
            "elapsedMs": Int(Date().timeIntervalSince(started) * 1_000),
          ])
        }
        terminal = true
        return [
          "requestId": parsed.requestId,
          "localUri": output.absoluteString,
          "mimeType": "image/png",
          "width": 512,
          "height": 512,
          "adapterId": "expo-sticker-runtime-coreml",
          "temporary": true,
        ]
      } catch {
        terminal = true
        let output = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0]
          .appendingPathComponent("generated-stickers/\(parsed.requestId).png")
        try? FileManager.default.removeItem(at: output)
        let code = error is GenerationCancellationError
          ? "GENERATION_CANCELLED"
          : (error as? StickerRuntimeError)?.code ?? "INFERENCE_FAILED"
        throw NSError(
          domain: code,
          code: 1,
          userInfo: [NSLocalizedDescriptionKey: "\(code): \(error.localizedDescription)"]
        )
      }
    }

    AsyncFunction("cancel") { (requestId: String) -> [String: Any] in
      self.generationLock.lock()
      let active = self.activeRequest
      self.generationLock.unlock()
      let accepted = active?.id == requestId && active?.cancellation.cancel() == true
      return [
        "accepted": accepted,
        "outcome": accepted ? "cancellation_requested" : "not_found",
      ]
    }
  }

  private func capabilities() -> [String: Any] {
    let simulator = ProcessInfo.processInfo.environment["SIMULATOR_DEVICE_NAME"] != nil
    let memoryMb = Int(ProcessInfo.processInfo.physicalMemory / 1_048_576)
    let decision = RuntimePolicy.evaluate(
      isSimulator: simulator,
      machineIdentifier: machineIdentifier(),
      operatingSystemMajor: ProcessInfo.processInfo.operatingSystemVersion.majorVersion,
      physicalMemoryMb: memoryMb
    )
    if decision.supported {
      return [
        "supported": true,
        "adapterId": "expo-sticker-runtime-coreml",
        "totalMemoryClassMb": memoryMb,
        "deviceKind": "physical",
        "architecture": decision.architecture,
        "availableDelegates": ["ANE", "GPU", "CPU"],
        "selectedDelegate": "ANE",
        "runtimeVersion": "ios-17",
      ]
    }
    return [
      "supported": false,
      "reasonCode": decision.reasonCode ?? "DEVICE_UNSUPPORTED",
      "deviceKind": simulator ? "emulator" : "physical",
      "architecture": decision.architecture,
    ]
  }

  private func parseGenerationRequest(_ request: [String: Any]) throws -> StickerGenerationRequest {
    guard request["contractVersion"] as? String == "1.0",
      request["stylePresetId"] as? String == "chibi",
      (request["outputWidth"] as? NSNumber)?.intValue == 512,
      (request["outputHeight"] as? NSNumber)?.intValue == 512,
      let requestId = request["requestId"] as? String,
      !requestId.isEmpty,
      let prompt = request["prompt"] as? String,
      !prompt.isEmpty,
      let seed = (request["seed"] as? NSNumber)?.int64Value
    else {
      throw StickerRuntimeError.invalidRequest
    }
    return StickerGenerationRequest(requestId: requestId, prompt: prompt, seed: seed)
  }

  private func register(_ id: String, cancellation: GenerationCancellation) -> Bool {
    generationLock.lock()
    defer { generationLock.unlock() }
    guard activeRequest == nil else { return false }
    activeRequest = (id, cancellation)
    return true
  }

  private func unregister(_ id: String) {
    generationLock.lock()
    defer { generationLock.unlock() }
    if activeRequest?.id == id { activeRequest = nil }
  }

  private func bundleState() -> [String: Any] {
    do {
      let manifest = try bundles.loadManifest()
      try bundles.recoverCompletedDownload(manifest)
      return state(bundles.isReady(manifest) ? "ready" : "missing", manifest: manifest)
    } catch {
      return failedState(error)
    }
  }

  private func failedState(_ error: Error) -> [String: Any] {
    let code = (error as? ModelBundleError)?.code ?? "MODEL_PROMOTION_FAILED"
    return state("failed", errorCode: code)
  }

  private func state(
    _ status: String,
    manifest: DistributionManifest? = nil,
    errorCode: String? = nil
  ) -> [String: Any] {
    let resolved = manifest ?? (try? bundles.loadManifest())
    var value: [String: Any] = [
      "status": status,
      "modelId": resolved?.modelId ?? "lcm-sd15-chibi",
      "modelVersion": resolved?.modelVersion ?? "1.0.1-coreml.1",
      "downloadedBytes": status == "ready" ? (resolved?.artifactBytes ?? 0) : 0,
      "totalBytes": resolved?.artifactBytes ?? 0,
    ]
    if let errorCode { value["errorCode"] = errorCode }
    return value
  }

  private func machineIdentifier() -> String {
    var systemInfo = utsname()
    uname(&systemInfo)
    return withUnsafePointer(to: &systemInfo.machine) {
      $0.withMemoryRebound(to: CChar.self, capacity: 1) { String(cString: $0) }
    }
  }
}
